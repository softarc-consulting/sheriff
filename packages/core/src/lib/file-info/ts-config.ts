import { CompilerOptions } from 'typescript';

export interface TsConfig {
  compilerOptions: CompilerOptions;
  extends: string;
}
