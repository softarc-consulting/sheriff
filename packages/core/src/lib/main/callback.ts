import { SheriffConfig } from '../config/sheriff-config';

export type Callback = (config: SheriffConfig | undefined) => void;
