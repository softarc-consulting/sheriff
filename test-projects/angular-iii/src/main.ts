import { importProvidersFrom, LOCALE_ID } from '@angular/core';

import { bootstrapApplication } from '@angular/platform-browser';
import { AppComponent } from './app/app.component';
import { Configuration } from '@eternal/shared/config';
import { MAT_DATE_LOCALE } from '@angular/material/core';
import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { BaseUrlInterceptor } from '@eternal/shared/http';
import {
  LoadingInterceptor,
  sharedUiMessagingProvider,
} from '@eternal/shared/ui-messaging';
import { MAT_FORM_FIELD_DEFAULT_OPTIONS } from '@angular/material/form-field';
import { provideAnimations } from '@angular/platform-browser/animations';
import { provideStore } from '@ngrx/store';
import { provideEffects } from '@ngrx/effects';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { securityProvider } from '@eternal/shared/security';
import { FormlyModule } from '@ngx-formly/core';
import { FormlyMaterialModule } from '@ngx-formly/material';
import { sharedMasterDataProvider } from '@eternal/shared/master-data';
import { appRoutes } from './app/app.routes';
import { provideRouter } from '@angular/router';
import { registerLocaleData } from '@angular/common';

import localeDeAt from '@angular/common/locales/de-AT';
import { deAT } from 'date-fns/locale';
import { environment } from './environments/environment';
import { MatDateFnsModule } from '@angular/material-date-fns-adapter';

registerLocaleData(localeDeAt, 'de-AT');

bootstrapApplication(AppComponent, {
  providers: [
    provideAnimations(),
    provideRouter(appRoutes),

    provideStore(),
    provideEffects([]),
    provideStoreDevtools(),

    ...securityProvider,
    sharedMasterDataProvider,

    importProvidersFrom(
      HttpClientModule,
      FormlyModule.forRoot({
        extras: { lazyRender: true },
        validationMessages: [
          {
            name: 'required',
            message: 'This field is mandatory',
          },
        ],
      }),
      FormlyMaterialModule,
      MatDateFnsModule
    ),
    sharedUiMessagingProvider,
    {
      provide: Configuration,
      useFactory: () => new Configuration(environment.baseUrl),
    },
    {
      provide: MAT_DATE_LOCALE,
      useValue: deAT,
    },
    { provide: HTTP_INTERCEPTORS, multi: true, useClass: BaseUrlInterceptor },
    { provide: HTTP_INTERCEPTORS, multi: true, useClass: LoadingInterceptor },
    { provide: LOCALE_ID, useValue: 'de-AT' },
    {
      provide: MAT_FORM_FIELD_DEFAULT_OPTIONS,
      useValue: { appearance: 'outline' },
    },
  ],
});
