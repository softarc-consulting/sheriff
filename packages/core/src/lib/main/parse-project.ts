import { FsPath } from '../file-info/fs-path';
import { FileInfo } from '../modules/file.info';
import { generateUnassignedFileInfo } from '../file-info/generate-unassigned-file-info';
import { getProjectDirsFromFileInfo } from '../modules/get-project-dirs-from-file-info';
import { findModulePaths } from '../modules/find-module-paths';
import { createModules } from '../modules/create-modules';
import { fillFileInfoMap } from '../modules/fill-file-info-map';
import throwIfNull from '../util/throw-if-null';
import { TsData } from '../file-info/ts-data';

export type ParsedResult = {
  fileInfo: FileInfo;
  modulePaths: Set<FsPath>;
  rootDir: FsPath;
  getFileInfo: (path: FsPath) => FileInfo;
};

export const parseProject = (
  entryFile: FsPath,
  traverse: boolean,
  tsData: TsData,
  fileContent?: string,
): ParsedResult => {
  const unassignedFileInfo = generateUnassignedFileInfo(
    entryFile,
    !traverse,
    tsData,
    fileContent,
  );
  const rootDir = tsData.rootDir;

  const projectDirs = getProjectDirsFromFileInfo(unassignedFileInfo, rootDir);

  const fileInfoMap: Map<FsPath, FileInfo> = new Map();
  const getFileInfo = (path: FsPath) =>
    throwIfNull(fileInfoMap.get(path), `cannot find FileInfo for ${path}`);

  const modulePaths = findModulePaths(projectDirs);
  const modules = createModules(
    unassignedFileInfo,
    modulePaths,
    rootDir,
    fileInfoMap,
    getFileInfo,
  );
  fillFileInfoMap(fileInfoMap, modules);

  const fileInfo = getFileInfo(unassignedFileInfo.path);

  return {
    fileInfo,
    rootDir,
    getFileInfo,
    modulePaths,
  };
};
