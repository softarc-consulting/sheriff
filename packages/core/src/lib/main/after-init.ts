import { SheriffConfig } from '../config/sheriff-config';
import { initialized } from './internal/initialized';
import { callbacks } from './internal/callback';

export function afterInit(
  callback: (config: SheriffConfig | undefined) => void,
) {
  if (initialized.status) {
    callback(initialized.config);
  } else {
    callbacks.push(callback);
  }
}
