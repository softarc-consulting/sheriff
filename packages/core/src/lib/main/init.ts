import { FsPath, toFsPath } from '../file-info/fs-path';
import { SheriffConfig } from '../config/sheriff-config';
import TsData from '../file-info/ts-data';
import getFs from '../fs/getFs';
import { generateTsData } from '../file-info/generate-ts-data';
import { findConfig } from '../config/find-config';
import { parseConfig } from '../config/parse-config';
import { ParsedResult, parseProject } from './parse-project';
import { initialized } from './internal/initialized';
import { callbacks } from './internal/callback';

let config: SheriffConfig | undefined;

export type ProjectInfo = {
  tsData: TsData;
  config: SheriffConfig | undefined;
} & ParsedResult;

/**
 * @property {boolean} [returnOnMissingConfig] - If config file is missing, would return undefined
 * @property {boolean} [traverse] - If import files should be traversed or stopped after entry file's
 * @property {string} [entryFileContent] - ESLint needed in order to process unsaved content
 */
export type InitOptions = {
  returnOnMissingConfig?: boolean;
  traverse?: boolean;
  entryFileContent?: string;
};

/**
 * Central initialisation which generates the central `FileInfo` element.
 * All tools (ESLint, CLI) need to call this one.
 *
 * It requires an entryFile `main.ts`. From there, it
 * locates the main `tsconfig.json` and defines its path as `rootPath`.
 *
 * The next step is to load `sheriff.config.ts`. The dependency rule checks
 * require a config. So there is an option, where the caller can
 * stop if the config is not available.
 *
 * Last step is to start with the analysis of the project by generating
 * the FileInfo. That it is a Tree structure, following the import commands
 * of the files where the root one is the entryFile.
 * For ESLint rules which focus on a single file, there is the option
 * to only include the imports of the entryFile but not traverse any
 * further.
 *
 * @param {FsPath} entryFile - The entry file path.
 * @param {InitOptions} options - option to traverse or return on no config
 *
 * @return {ProjectInfo} An object containing the generated TypeScript data and the configuration.
 **/
export function init(entryFile: FsPath): ProjectInfo;

export function init<Options extends { returnOnMissingConfig: true }>(
  entryFile: FsPath,
  options: InitOptions & Options
): ProjectInfo | undefined;

export function init(entryFile: FsPath, options: InitOptions): ProjectInfo;

export function init(entryFile: FsPath, options: InitOptions = {}) {
  const fullOptions = {
    ...{ traverse: true, returnOnMissingConfig: false },
    ...options,
  };
  const fs = getFs();
  const tsConfigPath = toFsPath(
    fs.findNearestParentFile(entryFile, 'tsconfig.json')
  );
  const tsData = generateTsData(tsConfigPath);
  config = getConfig(tsData.rootDir);
  if (config === undefined && fullOptions.returnOnMissingConfig) {
    return;
  }

  initialized.status = true;
  for (const callback of callbacks) {
    callback(config);
  }

  return {
    tsData,
    config,
    ...parseProject(
      entryFile,
      fullOptions.traverse,
      tsData,
      options.entryFileContent
    ),
  };
}

function getConfig(rootPath: FsPath) {
  const configFile = findConfig(rootPath);
  if (configFile) {
    return parseConfig(configFile);
  }

  return undefined;
}
