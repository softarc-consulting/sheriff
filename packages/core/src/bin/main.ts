#!/usr/bin/env node

/** TODO
 * - Show tags in list
 * - Add entryFile property in sheriff.config.ts
 * - Add documentation
 */

import * as process from 'process';
import { verify } from '../lib/cli/verify';
import { list } from '../lib/cli/list';
import { Cli } from '../lib/cli/util';
import { init } from '../lib/cli/init';

const [, , cmd, ...args] = process.argv;

const cli: Cli = {
  endProcessOk: () => process.exit(0),
  endProcessError: () => process.exit(1),
  log: (message: string) => console.log(message),
  logError: (message: string) => console.error(message),
};

switch (cmd) {
  case 'init':
    init(cli, args);
    break;
  case 'verify':
    verify(process.cwd(), cli, args);
    break;
  case 'list':
    list(args);
    break;
  default:
    console.log(
      '\u001b[1m' + 'Sheriff - Modularity for TypeScript Projects\x1b[0m',
    );
    console.log('');
    console.log('Commands:');
    console.log(
      '  sheriff init [main.ts]: initializes Sheriff by adding a sheriff.config.ts. If eslint is installed, it also adds the plugin',
    );
    console.log(
      '  sheriff list [main.ts]: lists the current modules of the project.',
    );
    console.log(
      '  sheriff verify [main.ts]: runs the verification process for the project',
    );

    break;
}
