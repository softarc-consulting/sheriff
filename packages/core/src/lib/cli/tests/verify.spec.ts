import { beforeEach, describe, expect, vitest, it } from 'vitest';
import { createProject } from '../../test/project-creator';
import { tsConfig } from '../../test/fixtures/ts-config';
import { main } from '../main';
import { sheriffConfig } from '../../test/project-configurator';
import { verifyCliWrappers } from './verify-cli-wrapper';
import { mockCli } from './helpers/mock-cli';

describe('verify', () => {
  beforeEach(() => {
    vitest.restoreAllMocks();
  });

  verifyCliWrappers('verify', 'src/main.ts');

  it('should find no errors', () => {
    const { allLogs, allErrorLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      src: {
        'main.ts': [''],
      },
    });

    main('verify', 'src/main.ts');

    expect(allErrorLogs()).toMatchSnapshot('error');
    expect(allLogs()).toMatchSnapshot('log');
  });

  it('should find errors', () => {
    const { allLogs, allErrorLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      'sheriff.config.ts': sheriffConfig({
        modules: {
          'src/customers': ['customers'],
          'src/holidays': ['holidays'],
        },
        depRules: {
          root: ['customers', 'holidays'],
          customers: [],
          holidays: [],
        },
      }),
      src: {
        'main.ts': ['./holidays', './customers', './customers/data'],
        holidays: {
          'index.ts': ['./holidays.component'],
          'holidays.component.ts': ['../customers'],
        },
        customers: { 'index.ts': [], 'data.ts': [] },
      },
    });

    main('verify', 'src/main.ts');

    expect(allErrorLogs()).toMatchSnapshot('error.log');
    expect(allLogs()).toMatchSnapshot('logs.log');
  });

  it('should find errors without sheriff.config.ts', () => {
    const { allLogs, allErrorLogs } = mockCli();

    createProject({
      'tsconfig.json': tsConfig(),
      src: {
        'main.ts': ['./holidays', './customers', './customers/data'],
        holidays: {
          'index.ts': ['./holidays.component'],
          'holidays.component.ts': ['../customers'],
        },
        customers: { 'index.ts': [], 'data.ts': [] },
      },
    });

    main('verify', 'src/main.ts');

    expect(allErrorLogs()).toMatchSnapshot('error.log');
    expect(allLogs()).toMatchSnapshot('logs.log');
  });

  describe('Multi project setup', () => {
    it('should find no errors when passing a single entryPoint', () => {
      const { allLogs, allErrorLogs } = mockCli();

      createProject({
        'tsconfig.json': tsConfig(),
        'sheriff.config.ts': sheriffConfig({
          depRules: {},
          entryPoints: {
            'project-i': 'projects/project-i/src/main.ts',
            'project-ii': 'projects/project-ii/src/main.ts',
          },
        }),
        projects: {
          'project-i': {
            src: {
              'main.ts': [],
              'app.ts': [],
            },
          },
          'project-ii': {
            src: {
              'main.ts': [],
              'app.ts': [],
            },
          },
        },
      });

      main('verify', 'project-i');

      expect(allErrorLogs()).toMatchSnapshot('error');
      expect(allLogs()).toMatchSnapshot('log');
    });
    it('should find no errors when passing multiple entryPoints', () => {
      const { allLogs, allErrorLogs } = mockCli();

      createProject({
        'tsconfig.json': tsConfig(),
        'sheriff.config.ts': sheriffConfig({
          depRules: {},
          entryPoints: {
            'project-i': 'projects/project-i/src/main.ts',
            'project-ii': 'projects/project-ii/src/main.ts',
          },
        }),
        projects: {
          'project-i': {
            src: {
              'main.ts': [],
              'app.ts': [],
            },
          },
          'project-ii': {
            src: {
              'main.ts': [],
              'app.ts': [],
            },
          },
        },
      });

      main('verify', 'project-i,project-ii');

      expect(allErrorLogs()).toMatchSnapshot('error');
      expect(allLogs()).toMatchSnapshot('log');
    });

    it('should find errors', () => {
      const { allLogs, allErrorLogs } = mockCli();

      createProject({
        'tsconfig.json': tsConfig(),
        'sheriff.config.ts': sheriffConfig({
          entryPoints: {
            'project-i': 'projects/project-i/src/main.ts',
            'project-ii': 'projects/project-ii/src/main.ts',
          },
          modules: {
            'projects/project-i': {
              'src/customers': ['customers'],
              'src/holidays': ['holidays'],
            },
          },
          depRules: {
            root: ['customers', 'holidays'],
            customers: [],
            holidays: [],
          },
        }),
        projects: {
          'project-i': {
            src: {
              'main.ts': ['./holidays', './customers', './customers/data'],
              holidays: {
                'index.ts': ['./holidays.component'],
                'holidays.component.ts': ['../customers'],
              },
              customers: { 'index.ts': [], 'data.ts': [] },
            },
          },
          'project-ii': {
            src: {
              'main.ts': [],
              'app.ts': [],
            },
          },
        },
      });

      main('verify');

      expect(allErrorLogs()).toMatchSnapshot('error.log');
      expect(allLogs()).toMatchSnapshot('logs.log');
    });
  });
});
