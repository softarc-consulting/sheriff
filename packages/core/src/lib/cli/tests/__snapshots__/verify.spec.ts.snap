// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`verify > should find errors > error.log 1`] = `""`;

exports[`verify > should find errors > logs.log 1`] = `
"
<b>Verification Report</b>

Issues found:
  Total Invalid Files: 2
  Total Deep Imports: 1
  Total Dependency Rule Violations: 1
----------------------------------

|-- src/main.ts
|   |-- Deep Imports
|   |   |-- ./customers/data
|-- src/holidays/holidays.component.ts
|   |-- Dependency Rule Violations
|   |   |-- from tags holidays to customers"
`;

exports[`verify > should find errors without sheriff.config.ts > error.log 1`] = `""`;

exports[`verify > should find errors without sheriff.config.ts > logs.log 1`] = `
"
<b>Verification Report</b>

Issues found:
  Total Invalid Files: 1
  Total Deep Imports: 1
  Total Dependency Rule Violations: 0
----------------------------------

|-- src/main.ts
|   |-- Deep Imports
|   |   |-- ./customers/data"
`;

exports[`verify > should find no errors > error 1`] = `""`;

exports[`verify > should find no errors > log 1`] = `
"
<b>Verification Report</b>

[32mNo issues found. Well done![0m"
`;
